internal/storagetools/transfer_handler.go

```diff
@@ -0,0 +1,295 @@
+		return nil, fmt.Errorf("can't list files in the target storage: %w", err)
+	}
+	sourceFiles, err := storage.ListFolderRecursivelyWithPrefix(h.source, h.cfg.Prefix)
+	if err != nil {
+		return nil, fmt.Errorf("can't list files in the source storage: %w", err)
+	}
+	tracelog.InfoLogger.Printf("Total files in the source storage: %d", len(sourceFiles))
+
+	missingFiles := make(map[string]storage.Object, len(sourceFiles))
+	for _, sourceFile := range sourceFiles {
+		missingFiles[sourceFile.GetName()] = sourceFile
+	}
+	for _, targetFile := range targetFiles {
+		if h.cfg.Overwrite {
+			sourceFile := missingFiles[targetFile.GetName()]
+			logSizesDifference(sourceFile, targetFile)
+		} else {
+			delete(missingFiles, targetFile.GetName())
+		}
+	}
+	tracelog.InfoLogger.Printf("Files missing in the target storage: %d", len(missingFiles))
+
+	count := 0
+	limitedFiles := make([]storage.Object, 0, utility.Min(h.cfg.MaxFiles, len(missingFiles)))
+	for _, file := range missingFiles {
+		if count >= h.cfg.MaxFiles {
+			break
+		}
+		limitedFiles = append(limitedFiles, file)
+		count++
+	}
+	tracelog.InfoLogger.Printf("Files will be transferred: %d", len(limitedFiles))
+
+	return limitedFiles, nil
+}
+
+func logSizesDifference(sourceFile, targetFile storage.Object) {
+	if sourceFile.GetSize() != targetFile.GetSize() {
+		tracelog.WarningLogger.Printf(
+			"File present in both storages and its size is different: %q (source %d bytes VS target %d bytes)",
+			targetFile.GetName(),
+			sourceFile.GetSize(),
+			targetFile.GetSize(),
+		)
+	}
+}
+
+type transferJob struct {
+	jobType         transferJobType
+	filePath        string
+	prevCheck       time.Time
+	performedChecks uint
+}
+
+type transferJobType int
+
+const (
+	transferJobTypeCopy transferJobType = iota
+	transferJobTypeDelete
+)
+
+func (h *TransferHandler) transferConcurrently(workers int, files []storage.Object) (finErr error) {
+	jobsQueue := make(chan transferJob, len(files))
+	for _, f := range files {
+		jobsQueue <- transferJob{
+			jobType:  transferJobTypeCopy,
+			filePath: f.GetName(),
+		}
+	}
+
+	errs := make(chan error, len(files))
+
+	workersCtx, cancelWorkers := context.WithCancel(context.Background())
+	workersWG := new(sync.WaitGroup)
+	workersWG.Add(workers)
+	for i := 0; i < workers; i++ {
+		go h.transferFilesWorker(workersCtx, jobsQueue, errs, workersWG)
+	}
+
+	errsWG := new(sync.WaitGroup)
+	errsWG.Add(1)
+	go func() {
+		defer errsWG.Done()
+		errsNum := 0
+		for e := range errs {
+			if h.cfg.FailOnFirstErr {
+				cancelWorkers()
+				finErr = e
+				break
+			}
+			tracelog.ErrorLogger.PrintError(e)
+			errsNum++
+			finErr = fmt.Errorf("finished with %d errors", errsNum)
+		}
+	}()
+
+	workersWG.Wait()
+	close(errs)
+	errsWG.Wait()
+
+	return finErr
+}
+
+func (h *TransferHandler) transferFilesWorker(
+	ctx context.Context,
+	jobsQueue chan transferJob,
+	errs chan error,
+	wg *sync.WaitGroup,
+) {
+	defer wg.Done()
+
+	for {
+		var job transferJob
+
+		select {
```

Comment:
Why we can't combine these two select statements into a single one?

Reply:
We can, but in case both channels are ready (`jobsQueue` is always ready), `select` chooses the random one. It means the command may transfer some more files even after the context has been cancelled. Although it's not a critical problem, it wasn't so hard for me to make it quit instantly